core/commands/add.go

```diff
@@ -265,110 +267,158 @@ You can now refer to the added file in a gateway, like so:
+
+			LOOP:
+				for {
+					select {
+					case out, ok := <-outChan:
+						if !ok {
+							if quieter {
+								fmt.Fprintln(os.Stdout, lastHash)
+							}
+
+							break LOOP
+						}
+						output := out.(*coreunix.AddedObject)
+						if len(output.Hash) > 0 {
+							lastHash = output.Hash
+							if quieter {
+								continue
+							}
+
+							if progress {
+								// clear progress bar line before we print "added x" output
+								fmt.Fprintf(os.Stderr, "\033[2K\r")
+							}
+							if quiet {
+								fmt.Fprintf(os.Stdout, "%s\n", output.Hash)
+							} else {
+								fmt.Fprintf(os.Stdout, "added %s %s\n", output.Hash, output.Name)
+							}
+
+						} else {
+							log.Debugf("add progress: %v %v\n", output.Name, output.Bytes)
+
+							if !progress {
+								continue
+							}
+
+							if len(lastFile) == 0 {
+								lastFile = output.Name
+							}
+							if output.Name != lastFile || output.Bytes < lastBytes {
+								prevFiles += lastBytes
+								lastFile = output.Name
+							}
+							lastBytes = output.Bytes
+							delta := prevFiles + lastBytes - totalProgress
+							totalProgress = bar.Add64(delta)
+						}
+
+						if progress {
+							bar.Update()
+						}
+					case size := <-sizeChan:
+						if progress {
+							bar.Total = size
+							bar.ShowPercent = true
+							bar.ShowBar = true
+							bar.ShowTimeLeft = true
+						}
+					case <-req.Context().Done():
+						re.SetError(req.Context().Err(), cmdsutil.ErrNormal)
+						return
 					}
-					break LOOP
 				}
-				output := out.(*coreunix.AddedObject)
-				if len(output.Hash) > 0 {
-					lastHash = output.Hash
-					if quieter {
-						continue
-					}
+			}
 
-					if progress {
-						// clear progress bar line before we print "added x" output
-						fmt.Fprintf(res.Stderr(), "\033[2K\r")
-					}
-					if quiet {
-						fmt.Fprintf(res.Stdout(), "%s\n", output.Hash)
-					} else {
-						fmt.Fprintf(res.Stdout(), "added %s %s\n", output.Hash, output.Name)
-					}
-				} else {
-					log.Debugf("add progress: %v %v\n", output.Name, output.Bytes)
+			go func() {
+				defer close(outChan)
+				defer re.Close()
 
-					if !progress {
-						continue
-					}
+				if e := res.Error(); e != nil {
+					re.SetError(e.Message, e.Code)
+					return
+				}
 
-					if len(lastFile) == 0 {
-						lastFile = output.Name
-					}
-					if output.Name != lastFile || output.Bytes < lastBytes {
-						prevFiles += lastBytes
-						lastFile = output.Name
+				go progressBar()
+
+				for {
+					v, err := res.Next()
+					log.Debug("add PostRun: next returned ", v, err)
+
+					if err != nil {
+						// replace error by actual error - will be looked at by next if-statement
+						if err == cmds.ErrRcvdError {
+							err = res.Error()
+						}
+
+						if e, ok := err.(*cmdsutil.Error); ok {
+							re.SetError(e.Message, e.Code)
+						}
+
+						// TODO why does err != io.EOF return true?
+						if err.Error() != "EOF" {
```

Comment:
odd. Is this still the case? Maybe we recreated an error somewhere like `errors.New(err.Error())`